### PR TITLE
Feature/protocol NAK reason param

### PR DESCRIPTION
**Pull request by ShobhitAd**
ATF Test Scripts to check the feature  [#3434](https://github.com/smartdevicelink/sdl_core/issues/3434)

This PR is **ready** for review.

### Summary
[Summary of PR changes]

### ATF version
develop

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/Protocol/NACKReason/001_Second_Protected_Video_Service_NACK.lua [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+--    a) Send a StartService ACK message
+--  7) Send another StartService Request to start a second protected VIDEO service
+--  SDL Does:
+--    a) Send a StartService NACK message with an empty bson payload(no reason param)
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local runner = require('user_modules/script_runner')
+local utils = require("user_modules/utils")
+if not utils.isFileExist("lib/bson4lua.so") then
+  runner.skipTest("'bson4lua' library is not available in ATF")
```
Comment (iCollin) [inline test_scripts/Protocol/NACKReason/001_Second_Protected_Video_Service_NACK.lua:32]: should this test skip or fail if there is no bson lib? I just wonder with the atf runner if we'd ever notice such an error if we just skip the test.

File: test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua [hunk 2]
```diff
@@ -0,0 +1,103 @@
+---------------------------------------------------------------------------------------------------
+--  Precondition:
+--  1) Initialize the client side certifcate file for SDL
+--  2) Start SDL, HMI, connect Mobile device
+--  3) Register App_1 and App_2 on SDL.
+--
+--  Steps:
```
Comment (iCollin) [inline test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua:7]: description here doesn't match the test

File: test_scripts/Protocol/commonProtocol.lua [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+function common.startServiceProtectedACK(pAppId, pServiceId, pRequestPayload, pResponsePayload)
+    local mobSession = common.getMobileSession(pAppId)
+    mobSession:StartSecureService(pServiceId, bson.to_bytes(pRequestPayload))
+    mobSession:ExpectControlMessage(pServiceId, {
+      frameInfo = common.frameInfo.START_SERVICE_ACK,
+      encryption = true
+    })
+    :ValidIf(function(_, data)
+        local actPayload = bson.to_table(data.binaryData)
+        -- utils.printTable(actPayload)
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply test_scripts/Protocol/commonProtocol.lua]: Fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2436/commits/5dd8216630c3968e4dc114a1059340f75dd157de

File: test_scripts/Protocol/commonProtocol.lua [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+function common.startServiceUnprotectedACK(pAppId, pServiceId, pRequestPayload, pResponsePayload)
+    local mobSession = common.getMobileSession(pAppId)
+    mobSession:StartService(pServiceId, bson.to_bytes(pRequestPayload))
+    mobSession:ExpectControlMessage(pServiceId, {
+      frameInfo = common.frameInfo.START_SERVICE_ACK,
+      encryption = false
+    })
+    :ValidIf(function(_, data)
+        local actPayload = bson.to_table(data.binaryData)
+        utils.printTable(actPayload)
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply test_scripts/Protocol/commonProtocol.lua]: Fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2436/commits/5dd8216630c3968e4dc114a1059340f75dd157de

File: test_scripts/Protocol/commonProtocol.lua [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+function common.startServiceProtectedNACK(pAppId, pServiceId, pRequestPayload, pResponsePayload)
+    local mobSession = common.getMobileSession(pAppId)
+    mobSession:StartSecureService(pServiceId, bson.to_bytes(pRequestPayload))
+    mobSession:ExpectControlMessage(pServiceId, {
+        frameInfo = common.frameInfo.START_SERVICE_NACK,
+        encryption = false
+    })
+    :ValidIf(function(_, data)
+        local actPayload = bson.to_table(data.binaryData)
+        utils.printTable(actPayload)
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply test_scripts/Protocol/commonProtocol.lua]: Fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2436/commits/5dd8216630c3968e4dc114a1059340f75dd157de

File: test_scripts/Protocol/commonProtocol.lua [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        return compareValues(pResponsePayload, actPayload, "binaryData")
+    end)
+end
+
+
+function common.startServiceUnprotectedNACK(pAppId, pServiceId, pRequestPayload, pResponsePayload)
+    print('In the NAK one')
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply test_scripts/Protocol/commonProtocol.lua]: Fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2436/commits/5dd8216630c3968e4dc114a1059340f75dd157de

File: test_scripts/Protocol/commonProtocol.lua [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    local mobSession = common.getMobileSession(pAppId)
+    mobSession:StartService(pServiceId, bson.to_bytes(pRequestPayload))
+    mobSession:ExpectControlMessage(pServiceId, {
+        frameInfo = common.frameInfo.START_SERVICE_NACK,
+        encryption = false
+    })
+    :ValidIf(function(_, data)
+        local actPayload = bson.to_table(data.binaryData)
+        utils.printTable(actPayload)
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply test_scripts/Protocol/commonProtocol.lua]: Fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2436/commits/5dd8216630c3968e4dc114a1059340f75dd157de

File: test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua (hunk 2 above)
Comment (ShobhitAd) [reply test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua:7]: Good catch. Fixed in https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2436/commits/5dd8216630c3968e4dc114a1059340f75dd157de

File: test_scripts/Protocol/NACKReason/001_Second_Protected_Video_Service_NACK.lua (hunk 1 above)
Comment (ShobhitAd) [reply test_scripts/Protocol/NACKReason/001_Second_Protected_Video_Service_NACK.lua:32]: I see the reasoning behind failing the tests if the bson_lib isn't there. I guess one possible scenario for the skipping the tests might be if we were purposely testing without the bson_lib, in which case skipping the tests would make more sense than marking them as failed tests. I'm fine with either approach, let me know what you think.

File: test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+--[[ Scenario ]]
+runner.Title("Preconditions")
+runner.Step("Clean environment", common.preconditions)
+runner.Step("Init SDL certificates", common.initSDLCertificates, { "./files/Security/client_credential_expired.pem"})
+runner.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
+runner.Step("Register App 1", common.registerAppWOPTU, {1})
+runner.Step("Register App 2", common.registerApp, {2})
+
+runner.Title("Test NAK reason param(protocol version 5.3.0)")
+runner.Step("PolicyTableUpdate", common.policyTableUpdate)
```
Comment (iCollin) [inline test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua:95]: Is the same `common.policyTableUpdate` call really toggling the certificate from missing to expired?

Comment (ShobhitAd) [reply test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua:97]: `common.policyTableUpdate` is starting the PTU but the actual update and sending the response is handled in the `missingCertificateNACK` and `expiredCertificateNACK` functions by `common.PolicyTableUpdateSuccess(ptUpdate)`

Comment (iCollin) [reply test_scripts/Protocol/NACKReason/004_Invalid_Certificate_NACK.lua:97]: of course. sorry I forgot that in re-review while comparing to the new description.

**Review by JackLivio — APPROVED** [review]

**Pull request merged**